kola/tests/kubeadm/templates.go

```diff
@@ -0,0 +1,384 @@
+                - name: nginx
+                  image: nginx
+                  ports:
+                  - containerPort: 80`
+
+	masterScript = `#!/bin/bash
+set -euo pipefail
+
+export RELEASE_VERSION={{ .ReleaseVersion }}
+export DOWNLOAD_DIR={{ .DownloadDir }}
+export PATH="${PATH}:${DOWNLOAD_DIR}"
+
+# create the required directory
+mkdir --parent \
+    /etc/systemd/system/kubelet.service.d \
+    ${HOME}/.kube \
+    /home/core/.kube
+
+# we download and install the various requirements:
+# * kubelet service and kubeadm dropin
+    
+curl --retry-delay 1 \
+    --retry 60 \
+    --retry-connrefused \
+    --retry-max-time 60 \
+    --connect-timeout 20 \
+    --fail \
+    -sSL \
+    "https://raw.githubusercontent.com/kubernetes/release/${RELEASE_VERSION}/cmd/kubepkg/templates/latest/deb/kubelet/lib/systemd/system/kubelet.service" |
+    sed "s:/usr/bin:${DOWNLOAD_DIR}:g" > /etc/systemd/system/kubelet.service
+    
+curl --retry-delay 1 \
+    --retry 60 \
+    --retry-connrefused \
+    --retry-max-time 60 \
+    --connect-timeout 20 \
+    --fail \
+    -sSL \
+    "https://raw.githubusercontent.com/kubernetes/release/${RELEASE_VERSION}/cmd/kubepkg/templates/latest/deb/kubeadm/10-kubeadm.conf" |
+    sed "s:/usr/bin:${DOWNLOAD_DIR}:g" > /etc/systemd/system/kubelet.service.d/10-kubeadm.conf
+    
+# we create the kubeadm config
+# plugin-volume-dir and flex-volume-plugin-dir are required since /usr is read-only mounted
+# etcd is also defined as external. The provided one has some issues with docker and selinux
+# (permission denied with /var/lib/etcd) so it can't boot properly
+cat << EOF > kubeadm-config.yaml
+apiVersion: kubeadm.k8s.io/v1beta2
+kind: InitConfiguration
+nodeRegistration:
+  kubeletExtraArgs:
+    volume-plugin-dir: "/opt/libexec/kubernetes/kubelet-plugins/volume/exec/"
+---
+apiVersion: kubeadm.k8s.io/v1beta2
+kind: ClusterConfiguration
+networking:
+  podSubnet: {{ .PodSubnet }}
+controllerManager:
+  extraArgs:
+    flex-volume-plugin-dir: "/opt/libexec/kubernetes/kubelet-plugins/volume/exec/"
+etcd:
+  external:
+    endpoints:
+    {{ range $endpoint := .Endpoints }}
+      - {{ $endpoint }}
+    {{ end }}
+EOF
+
+cat << EOF > calico.yaml
+# Source: https://docs.projectcalico.org/manifests/custom-resources.yaml
+apiVersion: operator.tigera.io/v1
+kind: Installation
+metadata:
+  name: default
+spec:
+  # Configures Calico networking.
+  calicoNetwork:
+    # Note: The ipPools section cannot be modified post-install.
+    ipPools:
+      - blockSize: 26
+        cidr: {{ .PodSubnet }}
+        encapsulation: VXLANCrossSubnet
+        natOutgoing: Enabled
+        nodeSelector: all()
+  flexVolumePath: /opt/libexec/kubernetes/kubelet-plugins/volume/exec/
+EOF
+
+{
+    systemctl enable --quiet --now kubelet
+    kubeadm config images pull
+    kubeadm init --config kubeadm-config.yaml
+    cp -i /etc/kubernetes/admin.conf $HOME/.kube/config
+    cp -i /etc/kubernetes/admin.conf /home/core/.kube/config
```

Comment:
These two lines are the same, or?

Reply:
`cp -i` means `--interactive` and is probably not the right behavior here?

Reply:
yeah it runs as root, I took the same way as the current `kubernetes` tests. Do you think it's better to run everything as a regular user and to invoke `sudo` when required ? I don't have a strong opinion on this test

Reply:
I think it's more normal to run most things as `core` and add `sudo` where necessary but if in this case we would need to add `sudo` add too many places it's not worth it and one could do it the other way round if wanted, e.g., `sudo -u core kubectl`.

Reply:
The `cp -i` is the only change I see which would be good to do before merging, the rest is not so important.